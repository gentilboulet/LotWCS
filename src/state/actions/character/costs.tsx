import { createAction } from "typesafe-actions";
import { ICost } from "../../models/character/costs";

export const payCost = createAction("character/COST_PAY", (cost: ICost) => {
  return { cost };
})();
