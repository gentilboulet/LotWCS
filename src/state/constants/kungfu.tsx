export const KUNGFU_OPEN_STYLE = 'KUNGFU_OPEN_STYLE';
export type KUNGFU_OPEN_STYLE = typeof KUNGFU_OPEN_STYLE;

export const KUNGFU_CUSTOM_NAME_FOR_STYLE = 'KUNGFU_CUSTOM_NAME_FOR_STYLE';
export type KUNGFU_CUSTOM_NAME_FOR_STYLE = typeof KUNGFU_CUSTOM_NAME_FOR_STYLE;

export const KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE = 'KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE';
export type KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE = typeof KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE;

export const KUNGFU_BUY_TECHNIQUE = 'KUNGFU_BUY_TECHNIQUE';
export type KUNGFU_BUY_TECHNIQUE = typeof KUNGFU_BUY_TECHNIQUE;
