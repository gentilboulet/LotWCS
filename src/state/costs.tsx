import { IDiscount, updateDiscounts } from 'state/discounts';
import { IStoreState } from 'state/type';


export interface ICost {
  destiny: number;
  discountIdx?: number;
  discountNewValue?: number;
  entanglement: number;
}

function _costFactory(state: IStoreState, idx: number, cost: number): ICost {
  if (idx === -1) {  return { destiny: cost, entanglement: 0 }; }
  else {
    const discount = state.discounts[idx].value;
    const remainingCost = Math.max(0, cost - discount);
    const usedDiscountValue = cost - remainingCost;
    return {
      destiny: remainingCost,
      discountIdx: idx,
      discountNewValue: discount - usedDiscountValue,
      entanglement: 0,
    };
  }
}

export const emptyCost: ICost = { destiny: 0, entanglement: 0 };

export function canPayCost(state:IStoreState, cost: ICost): boolean {
  if(state.destiny < cost.destiny ) { return false; }
  if(state.entanglement < cost.entanglement ) { return false; }
  return true;
}

export function applyCost(state: IStoreState, cost: ICost): void {
    state.destiny -= cost.destiny;
    state.entanglement -= cost.entanglement;

    if (state.destiny < 0) { throw new Error('Negative destiny reached'); }
    if (state.entanglement < 0) { throw new Error('Negative entanglement reached'); }
    updateDiscounts(state, cost);

}

import { TSkillName } from 'data/skills';
import { DISCOUNT_SKILL } from 'state/constants/perks/discounts';
export function getCostSkill(state: IStoreState, skillName: TSkillName): ICost {
  const defaultCost = 2;

  const idx = state.discounts.findIndex((d:IDiscount) => {
    if(d.type !== DISCOUNT_SKILL) { return false }
    if(d.skills.length === 0) { return true; } // Any skills
    return -1 !== d.skills.findIndex((s: TSkillName) => s === skillName);
  })

  return _costFactory(state, idx, defaultCost);
}

export function getCostSpeciality(state: IStoreState, skillName: TSkillName, speciality: string): ICost {
  const defaultCost = 1;

  const idx = state.discounts.findIndex((d:IDiscount) => {
    if(d.type !== DISCOUNT_SKILL) { return false }
    if(d.skills.length === 0) { return true; } // Any skills
    return -1 !== d.skills.findIndex((s: TSkillName) => s === skillName);
  })

  return _costFactory(state, idx, defaultCost);
}

import { TChiName } from 'data/chi';
import { DISCOUNT_CHI } from 'state/constants/perks/discounts';
export function getCostChi(state: IStoreState, chiName: TChiName): ICost {
  const defaultCost = state.chi[chiName].value; // TODO cultivation

  const idx = state.discounts.findIndex((d:IDiscount) => {
    if(d.type !== DISCOUNT_CHI) { return false }
    if(d.chis.length === 0) { return true; } // Any chi
    return -1 !== d.chis.findIndex((s: TChiName) => s === chiName);
  })

  return _costFactory(state, idx, defaultCost);
}

export function getCostOpenLoresheet(state: IStoreState, uid: string): ICost {
/*  const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_LORESHEET )
      && (r.uids.findIndex((id: string) => (id === uid)) >= 0);
    });
  return _handleDiscount(state, idx, cost);
*/
  return _costFactory(state, -1, 1);
}

export function getCostBuyLoresheetOption(state: IStoreState, lsUid: string, uid: string): ICost {
  /*const idx = state.get('discounts')
    .findIndex((r: IDiscount) => {
      return ( r.type === constants.DISCOUNT_LORESHEET_OPTION )
      && (r.uids.findIndex(id => { return id.lsUid === lsUid &&
        id.optUid.findIndex(optUid => (optUid === uid)) >= 0; })
      >= 0);
    });
  return _handleDiscount(state, idx, cost);*/
  return _costFactory(state, -1, 1);
}

/*


export function canBuyOptionLoresheet(state: IStoreState, lsUid: string, uid: string, buyCost: number): boolean {
  if ( loresheets.getLoresheetIndex(state, lsUid) === -1) { return false; } // LS not open

  const dataOpt = dataLoresheets.optionLS(lsUid, uid);
  const idxOpt = loresheets.getLoresheetOptionIndex(state, lsUid, uid);

  if (idxOpt === -1 && !dataOpt.repeatable) { return false; } // Already bought & not repeatable

  if (dataOpt.prerequisites.filter( // filter checked prereqs
      (p: dataLoresheets.IDataLoresheetOptionPrerequisite) => {
        if (typeof p === 'string') {
          return loresheets.getLoresheetOptionIndex(state, lsUid, p) !== -1;
        } else if ( p.type === 'OR' ) {
          return p.prerequisites
            .filter(
              (prerequisiteUid: string) => (loresheets.getLoresheetOptionIndex(state, lsUid, prerequisiteUid) !== -1)
            ).length > 0;
        } else if ( p.type === 'AND') {
          return p.prerequisites
            .filter(
              (prerequisiteUid: string) => (loresheets.getLoresheetOptionIndex(state, lsUid, prerequisiteUid) !== -1)
            ).length === p.prerequisites.length;
        }
        return false;
      }
    ).length !== dataOpt.prerequisites.length ) { return false; } // prereqs not cleared

  const cost: ICost = getCostBuyOptionLoresheet(state, lsUid, uid, buyCost);
  return _canHandleCost(state, cost);
}

*/

export function getCostVirtue(state: IStoreState, name: string): ICost {
  /*
   DISCOUNT !

  */
  return _costFactory(state, -1, 5);
}

export function getCostKungFuStyle(state: IStoreState, type: any, uid: string): ICost {
  /*
   DISCOUNT !
  */
  return _costFactory(state, -1, 10);
}

export function getCostKungFuTechnique(state: IStoreState, type: any, styleUid: string, uid: string): ICost {
  /*
   DISCOUNT !
  */
  return _costFactory(state, -1, 10);
}
