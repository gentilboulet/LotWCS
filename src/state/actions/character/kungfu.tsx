import { createAction } from "typesafe-actions";

import * as dataKungFu from "../../../data/kungfu";
import { KUNGFU_TYPE } from "../../../data/kungfu/types";
import { bonusCultivation } from "../../../perks/actions/bonuses";
import { ICost } from "../../models/character/costs";
import { historyMetaCreator } from "../meta";

export const openStyle = createAction(
  "kungfu/OPEN_STYLE",
  (uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);

    return {
      cost,
      uid,
      kungfuType,
      perks: [
        bonusCultivation(
          dataKungFu.getKungfuChi(kungfuType, uid),
          cost.original,
        ),
      ],
    };
  },
  historyMetaCreator,
)();

export const buyTechnique = createAction(
  "kungfu/BUY_TECHNIQUE",
  (styleUid: string, uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    return {
      cost,
      kungfuType,
      styleUid,
      uid,
      perks: [
        bonusCultivation(
          dataKungFu.getKungfuChi(kungfuType, styleUid),
          cost.original,
        ),
      ],
    };
  },
  historyMetaCreator,
)();

export const customStyleName = createAction(
  "kungfu/CUSTOM_NAME_FOR_STYLE",
  (uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return {
      kungfuType,
      name,
      uid,
    };
  },
  historyMetaCreator,
)();

export const customTechniqueName = createAction(
  "kungfu/CUSTOM_NAME_FOR_TECHNIQUE",
  (styleUid: string, uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return {
      kungfuType,
      name,
      styleUid,
      uid,
    };
  },
  historyMetaCreator,
)();
