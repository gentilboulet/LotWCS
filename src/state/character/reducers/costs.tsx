import { produce, Draft } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/costs";
import { applyCost } from "../models/costs";
import { ICharacterState } from "../models/type";

export type ICostAction = ActionType<typeof actions>;

export const costReducer = produce(
  (draft: Draft<ICharacterState>, action: ICostAction) => {
    switch (action.type) {
      case getType(actions.payCost):
        applyCost(draft, action.payload.cost);
        break;
      default:
    }
  },
);
