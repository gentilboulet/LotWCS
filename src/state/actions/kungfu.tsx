import * as dataKungFu from 'data/kungfu';
import * as constants from 'state/constants/kungfu';
import { ICost } from 'state/costs';

export interface IKungFuOpenStyle {
  type: constants.KUNGFU_OPEN_STYLE;
  cost: ICost;
  uid: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export interface IKungFuBuyTechnique {
  type: constants.KUNGFU_BUY_TECHNIQUE;
  cost: ICost;
  styleUid: string;
  uid: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export interface IKungFuSetCustomStyleName {
  type: constants.KUNGFU_CUSTOM_NAME_FOR_STYLE;
  uid: string;
  name: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export interface IKungFuSetCustomTechniqueName {
  type: constants.KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE;
  styleUid: string;
  uid: string;
  name: string;
  kungfuType: dataKungFu.KUNGFU_TYPE;
}

export type IKungFuAction = IKungFuOpenStyle | IKungFuBuyTechnique
| IKungFuSetCustomStyleName | IKungFuSetCustomTechniqueName

export function openStyle(uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, cost: ICost): IKungFuOpenStyle {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, uid);
  return { type: constants.KUNGFU_OPEN_STYLE, cost, uid, kungfuType }
}

export function buyTechnique(
  styleUid: string, uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, cost: ICost): IKungFuBuyTechnique {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, styleUid);
  dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
  return { type: constants.KUNGFU_BUY_TECHNIQUE, cost, styleUid, uid, kungfuType };
}

export function customStyleName(uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, name: string): IKungFuSetCustomStyleName {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, uid);
  if (name.length < 1) { throw new Error ('Invalid new name'); };
  return { type: constants.KUNGFU_CUSTOM_NAME_FOR_STYLE, uid, name , kungfuType }
}

export function customTechniqueName(
  styleUid: string, uid: string, kungfuType: dataKungFu.KUNGFU_TYPE, name: string
): IKungFuSetCustomTechniqueName {
  dataKungFu.validateKungFuType(kungfuType);
  dataKungFu.validateKungFuStyle(kungfuType, styleUid);
  dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
  if (name.length < 1) { throw new Error ('Invalid new name'); };
  return { type: constants.KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE, styleUid, uid, name , kungfuType }
}
