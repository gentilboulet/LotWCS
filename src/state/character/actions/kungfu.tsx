import { createAction } from "typesafe-actions";

import * as dataKungFu from "../../data/kungfu";
import { KUNGFU_TYPE } from "../../data/kungfu/types";
import { ICost } from "../costs";

export const openStyle = createAction(
  "kungfu/OPEN_STYLE",
  action => (uid: string, kungfuType: KUNGFU_TYPE, cost: ICost) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    return action({ uid, kungfuType, cost });
  }
);

export const buyTechnique = createAction(
  "kungfu/BUY_TECHNIQUE",
  action => (
    styleUid: string,
    uid: string,
    kungfuType: KUNGFU_TYPE,
    cost: ICost
  ) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    return action({
      cost,
      kungfuType,
      styleUid,
      uid
    });
  }
);

export const customStyleName = createAction(
  "kungfu/CUSTOM_NAME_FOR_STYLE",
  action => (uid: string, kungfuType: KUNGFU_TYPE, name: string) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return action({
      kungfuType,
      name,
      uid
    });
  }
);

export const customTechniqueName = createAction(
  "kungfu/CUSTOM_NAME_FOR_TECHNIQUE",
  action => (
    styleUid: string,
    uid: string,
    kungfuType: KUNGFU_TYPE,
    name: string
  ) => {
    dataKungFu.validateKungFuType(kungfuType);
    dataKungFu.validateKungFuStyle(kungfuType, styleUid);
    dataKungFu.validateKungFuTechnique(kungfuType, styleUid, uid);
    if (name.length < 1) {
      throw new Error("Invalid new name");
    }
    return action({
      kungfuType,
      name,
      styleUid,
      uid
    });
  }
);
