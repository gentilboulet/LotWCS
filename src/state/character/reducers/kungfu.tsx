import { produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../actions/kungfu";
import { ICharacterState } from "../models/type";

import { applyCost } from "../models/costs";

import * as dataKungfu from "../../../data/kungfu";
import * as chi from "../models/chi";
import { addKungFuTechnique, openStyle } from "../models/kungfu";

export function kungfuReducer(
  baseState: ICharacterState,
  action: ActionType<typeof actions>
): ICharacterState {
  switch (action.type) {
    case getType(actions.openStyle):
      return produce(baseState, draftState => {
        applyCost(draftState, action.payload.cost);
        openStyle(
          draftState.kungfu,
          action.payload.kungfuType,
          action.payload.uid
        );

        chi.increaseCultivation(
          draftState.chi,
          dataKungfu.getKungfuChi(
            action.payload.kungfuType,
            action.payload.uid
          ),
          action.payload.cost.original
        );
      });
    case getType(actions.buyTechnique):
      return produce(baseState, draftState => {
        applyCost(draftState, action.payload.cost);
        addKungFuTechnique(
          draftState.kungfu,
          action.payload.kungfuType,
          action.payload.styleUid,
          action.payload.uid
        );

        chi.increaseCultivation(
          draftState.chi,
          dataKungfu.getKungfuChi(
            action.payload.kungfuType,
            action.payload.styleUid
          ),

          action.payload.cost.original
        );
      });
    case getType(actions.customStyleName):
    case getType(actions.customTechniqueName):
    default:
  }
  return baseState;
}
