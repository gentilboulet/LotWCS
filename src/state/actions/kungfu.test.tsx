import * as dataKungfus from "../../data/kungfu";
import {
  KUNGFU_EXTERNAL,
  KUNGFU_INTERNAL,
  KUNGFU_TYPE
} from "../../data/kungfu/types";
import { zeroCost } from "../costs";
import * as actions from "./kungfu";

describe("Testing external kungfu action creator", () => {
  expect(dataKungfus.externalKungfu.length).toBeGreaterThan(0);
  expect(dataKungfus.internalKungfu.length).toBeGreaterThan(0);

  // Typescript error checking can't be sure arrays are not empty
  if (
    dataKungfus.externalKungfu.length === 0 ||
    dataKungfus.internalKungfu.length === 0
  ) {
    throw new Error("Impossible failure in tests");
  }

  const validExternal = dataKungfus.externalKungfu[0];
  const validInternal = dataKungfus.internalKungfu[0];

  it("should validate testing data :)", () => {
    expect(validExternal.uid.length).toBeGreaterThan(0);
    expect(validExternal.techniques.length).toBeGreaterThan(0);
    expect(validInternal.uid.length).toBeGreaterThan(0);
    expect(validInternal.techniques.length).toBeGreaterThan(0);
  });

  it("should create an OPEN_STYLE action with an internal kungfu", () => {
    const uid = validInternal.uid;
    const type = KUNGFU_INTERNAL;
    const a = actions.openStyle(uid, type, zeroCost);
    expect(a.uid).toBe(uid);
    expect(a.kungfuType).toBe(type);
    expect(a).toMatchSnapshot();
  });

  it("should create an OPEN_STYLE action with an external kungfu", () => {
    const uid = validExternal.uid;
    const type = KUNGFU_EXTERNAL;
    const a = actions.openStyle(uid, type, zeroCost);
    expect(a.uid).toBe(uid);
    expect(a.kungfuType).toBe(type);
    expect(a).toMatchSnapshot();
  });

  it("should not create an OPEN_STYLE action with an erroneous kungfu type", () => {
    const type = "TOTALLY A KUNGFU TYPE" as KUNGFU_TYPE;
    expect(() =>
      actions.openStyle(validExternal.uid, type, zeroCost)
    ).toThrow();
  });

  it("should not create an OPEN_STYLE action with an erroneous internal kungfu", () => {
    const uid = "TOTALLY A KUNGFU UID";
    expect(() => actions.openStyle(uid, KUNGFU_INTERNAL, zeroCost)).toThrow();
    expect(() => actions.openStyle("", KUNGFU_INTERNAL, zeroCost)).toThrow();
  });

  it("should not create an OPEN_STYLE action with an erroneous external kungfu", () => {
    const uid = "TOTALLY A KUNGFU UID";
    expect(() => actions.openStyle(uid, KUNGFU_EXTERNAL, zeroCost)).toThrow();
    expect(() => actions.openStyle("", KUNGFU_EXTERNAL, zeroCost)).toThrow();
  });

  it("should create a BUY_TECHNIQUE action with an internal kungfu", () => {
    const techUid = validInternal.techniques[0].uid;
    const uid = validInternal.uid;
    const type = KUNGFU_INTERNAL;
    const a = actions.buyTechnique(uid, techUid, type, zeroCost);
    expect(a.uid).toBe(techUid);
    expect(a.kungfuType).toBe(type);
    expect(a.styleUid).toBe(uid);
  });

  it("should create a BUY_TECHNIQUE action with an external kungfu", () => {
    const techUid = validExternal.techniques[0].uid;
    const uid = validExternal.uid;
    const type = KUNGFU_EXTERNAL;
    const a = actions.buyTechnique(uid, techUid, type, zeroCost);
    expect(a.uid).toBe(techUid);
    expect(a.kungfuType).toBe(type);
    expect(a.styleUid).toBe(uid);
  });

  it("should not create an BUY_TECHNIQUE action with an erroneous kungfu type", () => {
    const type = "TOTALLY A KUNGFU TYPE" as KUNGFU_TYPE;
    const techUid = validExternal.techniques[0].uid;
    expect(() =>
      actions.buyTechnique(techUid, validExternal.uid, type, zeroCost)
    ).toThrow();
  });

  it("should not create an BUY_TECHNIQUE action with an erroneous internal kungfu", () => {
    const uid = "TOTALLY A KUNGFU UID";
    const techUid = validInternal.techniques[0].uid;
    expect(() =>
      actions.buyTechnique(techUid, uid, KUNGFU_INTERNAL, zeroCost)
    ).toThrow();
    expect(() =>
      actions.buyTechnique(techUid, "", KUNGFU_INTERNAL, zeroCost)
    ).toThrow();
  });

  it("should not create an BUY_TECHNIQUE action with an erroneous external kungfu", () => {
    const uid = "TOTALLY A KUNGFU UID";
    const techUid = validExternal.techniques[0].uid;
    expect(() =>
      actions.buyTechnique(techUid, uid, KUNGFU_EXTERNAL, zeroCost)
    ).toThrow();
    expect(() =>
      actions.buyTechnique(techUid, "", KUNGFU_EXTERNAL, zeroCost)
    ).toThrow();
  });

  it("should not create an BUY_TECHNIQUE action with an erroneous internal kungfu technique", () => {
    const uid = validInternal.uid;
    const techUid = "TOTALLY A KUNGFU TECHNIQUE UID";
    expect(() =>
      actions.buyTechnique(techUid, uid, KUNGFU_INTERNAL, zeroCost)
    ).toThrow();
    expect(() =>
      actions.buyTechnique("", uid, KUNGFU_INTERNAL, zeroCost)
    ).toThrow();
  });

  it("should not create an BUY_TECHNIQUE action with an erroneous external kungfu technique", () => {
    const uid = validExternal.uid;
    const techUid = "TOTALLY A KUNGFU TECHNIQUE UID";
    expect(() =>
      actions.buyTechnique(techUid, uid, KUNGFU_EXTERNAL, zeroCost)
    ).toThrow();
    expect(() =>
      actions.buyTechnique("", uid, KUNGFU_EXTERNAL, zeroCost)
    ).toThrow();
  });
});

describe("Testing custom naming action for kungfu style and techniques", () => {
  const validExternal = dataKungfus.externalKungfu[0];

  it("should create a KUNGFU_CUSTOM_NAME_FOR_STYLE action for a valid style", () => {
    const uid = validExternal.uid;
    expect(
      actions.customStyleName(uid, KUNGFU_EXTERNAL, "new name")
    ).toMatchSnapshot();
  });

  it("should refuse to create a KUNGFU_CUSTOM_NAME_FOR_STYLE action for an invalid style", () => {
    const uid = "TOTALLY A KUNGFU UID";
    expect(() => {
      actions.customStyleName(uid, KUNGFU_EXTERNAL, "new name");
    }).toThrowError();
  });

  it("should refuse to create a KUNGFU_CUSTOM_NAME_FOR_STYLE action with an empty new name", () => {
    const uid = validExternal.uid;
    expect(() => {
      actions.customStyleName(uid, KUNGFU_EXTERNAL, "");
    }).toThrowError();
  });

  it("should create a KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE action for a valid style", () => {
    const uid = validExternal.uid;
    const techUid = validExternal.techniques[0].uid;

    expect(
      actions.customTechniqueName(uid, techUid, KUNGFU_EXTERNAL, "new name")
    ).toMatchSnapshot();
  });

  it("should refuse to create a KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE action for an invalid style", () => {
    const uid = "TOTALLY A KUNGFU UID";

    const techUid = validExternal.techniques[0].uid;
    expect(() => {
      actions.customTechniqueName(uid, techUid, KUNGFU_EXTERNAL, "new name");
    }).toThrowError();
  });

  it("should refuse to create a KUNGFU_CUSTOM_NAME_FOR_STYLE action for an invalid style", () => {
    const uid = validExternal.uid;
    const techUid = "TOTALLY A KUNGFU TECHNIQUE UID";
    expect(() => {
      actions.customTechniqueName(uid, techUid, KUNGFU_EXTERNAL, "new name");
    }).toThrowError();
  });

  it("should refuse to create a KUNGFU_CUSTOM_NAME_FOR_TECHNIQUE action with an empty new name", () => {
    const uid = validExternal.uid;
    const techUid = validExternal.techniques[0].uid;

    expect(() => {
      actions.customTechniqueName(uid, techUid, KUNGFU_EXTERNAL, "");
    }).toThrowError();
  });
});
