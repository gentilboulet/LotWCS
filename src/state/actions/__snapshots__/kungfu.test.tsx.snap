// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing external kungfu action creator should create an OPEN_STYLE action with an external kungfu 1`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "kungfuType": "KUNGFU_EXTERNAL",
  "type": "KUNGFU_OPEN_STYLE",
  "uid": "Blossom Harvest",
}
`;

exports[`Testing external kungfu action creator should create an OPEN_STYLE action with an internal kungfu 1`] = `
Object {
  "cost": Object {
    "destiny": 0,
    "discountIdx": -1,
    "discountNewValue": 0,
    "entanglement": 0,
  },
  "kungfuType": "KUNGFU_INTERNAL",
  "type": "KUNGFU_OPEN_STYLE",
  "uid": "Boundless Prosperity Manual",
}
`;
